File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem10_JobScheduler.scala
package dcp.problems._1_20

import java.util.{Timer, TimerTask}

/**
 * This problem was asked by Apple.
 *
 * Implement a job scheduler which takes in a function f and an integer n, and calls f after n milliseconds.
 */

object Problem10_JobScheduler {

  def execute(n: Long)(f: => Unit): Unit = {
    val timer = new Timer()
    val task = new TimerTask {
      override def run(): Unit = {
        f
        timer.cancel()
      }
    }
    timer.schedule(task, n)
  }

  def main(args: Array[String]): Unit = {
    val delay = execute(3000) _
    delay {
      println("Hello, did I win? :)")
    }
    execute(2000) {
      println("I am going to be first!")
    }
  }
}

File: daily-coding-problem/dcp-py/1_20/problem10_job_scheduler.py
"""
This problem was asked by Apple.

Implement a job scheduler which takes in a function f and an integer n, and calls f after n milliseconds.
"""
import time


def function():
    print("Executing function...")


def schedule(fun, delay):
    delay_in_seconds = delay / 1000
    print(f"Function will get executed after {delay_in_seconds} seconds...")
    time.sleep(delay_in_seconds)
    fun()


schedule(function, 3000)
